src/pe/certificate_table.rs

```diff
@@ -0,0 +1,146 @@
+    Revision1_0 = 0x0100,
+    Revision2_0 = 0x0200,
+}
+
+impl TryFrom<u16> for CertificateAttributeRevision {
+    type Error = error::Error;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        Ok(match value {
+            x if x == CertificateAttributeRevision::Revision1_0 as u16 => {
+                CertificateAttributeRevision::Revision1_0
+            }
+            x if x == CertificateAttributeRevision::Revision2_0 as u16 => {
+                CertificateAttributeRevision::Revision2_0
+            }
+            _ => {
+                return Err(error::Error::Malformed(
+                    "Invalid certificate attribute revision".to_string(),
+                ))
+            }
+        })
+    }
+}
+
+#[repr(u16)]
+#[derive(Debug)]
+enum CertificateAttributeType {
+    X509 = 0x0001,
+    PkcsSignedData = 0x0002,
+    Reserved1 = 0x0003,
+    TsStackSigned = 0x0004,
+}
+
+impl TryFrom<u16> for CertificateAttributeType {
+    type Error = error::Error;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        Ok(match value {
+            x if x == CertificateAttributeType::X509 as u16 => CertificateAttributeType::X509,
+            x if x == CertificateAttributeType::PkcsSignedData as u16 => {
+                CertificateAttributeType::PkcsSignedData
+            }
+            x if x == CertificateAttributeType::Reserved1 as u16 => {
+                CertificateAttributeType::Reserved1
+            }
+            x if x == CertificateAttributeType::TsStackSigned as u16 => {
+                CertificateAttributeType::TsStackSigned
+            }
+            _ => {
+                return Err(error::Error::Malformed(
+                    "Invalid certificate attribute type".to_string(),
+                ))
+            }
+        })
+    }
+}
+
+#[derive(Clone, Pread)]
+pub struct CertificateAttributeHeader {
+    length: u32,
+    revision: u16,
+    certificate_type: u16,
+}
+
+pub const CERTIFICATE_DATA_OFFSET: u32 = 8;
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct CertificateAttribute<'a> {
+    length: u32,
+    revision: CertificateAttributeRevision,
+    certificate_type: CertificateAttributeType,
+    certificate: &'a [u8],
+}
+
+impl<'a> CertificateAttribute<'a> {
+    pub fn parse(
+        bytes: &'a [u8],
+        current_offset: &mut usize,
+    ) -> Result<CertificateAttribute<'a>, error::Error> {
+        // `current_offset` is moved sizeof(CertificateAttributeHeader) = 8 bytes further.
+        let header: CertificateAttributeHeader = bytes.gread_with(current_offset, scroll::LE)?;
+        let mut cert_size = usize::try_from(header.length.saturating_sub(CERTIFICATE_DATA_OFFSET))
+            .map_err(|_err| {
+                error::Error::Malformed(
+                    "Certificate attribute size do not fit in usize".to_string(),
+                )
+            })?;
+        // Round to the next 8-bytes.
+        cert_size = (cert_size + 7) & !7;
```

Comment:
This is weird to me. This would affect the `certificate` field, and this should not be rounded yet, this should be rounded further down (before current_offset is bumped).
That would pad zero bytes (or garbage) in the certificate field otherwise.

Reply:
you're totally right